### PR TITLE
Documentation generator

### PR DESCRIPTION
**Pull request by bsuwal**
This only looks like a big PR - but it really isn't! i just ported all the documentation to be generated by Documenter.jl.

Some snippets of what it looks like:
Screen Shot 2020-10-14 at 2 49 07 PM

Screen Shot 2020-10-14 at 2 21 05 PM

The actual documentation pages only render once pushed to `main`, so we can't see them yet unless you look at them locally. To do so, do the following after pulling this branch.

The directory structure of the package is 

GerryChainJulia/
|- Project.toml
|- src/
|      |- ...
|- test/
|      |- ...
|- docs/
|      |- make.jl
|      |- ...
...

From the root of this repository, on the command-line, type `julia --project=. docs/make.jl`

This takes a few seconds and should build the HTML files for the documentation.

Next, do `python -m http.server --bind localhost` and go to whatever port it starts hosting from on your fav browser, go to `docs/` to view the documentation.

***Code changes***
I have not made any code changes. However I have changed all the docstrings to be of a format that the `Document.jl` renderer wants. Also, I have made a stylistic change for consistency, where I put each argument in a function signature in a different line for consistency purposes. I find it easier to read this way on the documentation pages and development in general, and thought I might as well make it consistent across the board. Thoughts?

More explicitly, functions that look like
```
function my_func(x::Int, y::Int)
    x + y
end
```
now look like 
```
function my_func(x::Int, 
                 y::Int)
    x + y
end
```

Apart from that, the changes I made were mostly to create the new `.md` files needed by `Documenter.jl` and then on the config files like `.travis.yml` and `.gitignore`. I ironed out some documentation that seemed outdated, and formatted some to render better in a Markdown for `Documenter.jl` otherwise the documentation has also mostly stayed the same.

### CODE REVIEW

File: docs/src/index.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+## _Why_ is GerryChain in Julia?
+
+At a fundamental level, GerryChain exists because it is infeasible to compare an initial plan to the complete universe of possible districting plans for a particular region. (In many cases, there are more possible plans than atoms in the universe!) Through the GerryChain approach, we instead take a "random walk" in the universe of possible plans, generating some finite number of plans, to which we can then compare our initial plan. Speeding up the GerryChain computations with Julia means that we can generate a number of plans an order of magnitude greater than what we could achieve with the Python library in the same amount of time, with the intention that the subsequent analysis is more robust.
+
+## Citations
+
+## Citations
```
Comment (matthewkol186) [inline docs/src/index.md:30]: Is this duplicated?

Comment (matthewkol186) [inline docs/src/index.md:30]: Is this duplicated?

File: docs/src/index.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    archivePrefix={arXiv},
+    primaryClass={cs.CY}
+}
+```
+
+The most up to date citation of the code would be at [![DOI](https://zenodo.org/badge/239854101.svg)](https://zenodo.org/badge/latestdoi/239854101).
+
+## How do I get started?
+
+You'll want to start with [setting up your computing environment](installation.md). Next, you should check out [our tutorial on running your first GerryChain](getting_started.md)!. If you wish to contribute to the development of this project, please refer to our [contributing guidelines](https://github.com/mggg/GerryChainJulia/blob/main/CONTRIBUTING.md).
```
Comment (matthewkol186) [inline docs/src/index.md:47]:
```suggestion
You'll want to start with [setting up your computing environment](installation.md). Next, you should check out [our tutorial on running your first GerryChain](getting_started.md)! If you wish to contribute to the development of this project, please refer to our [contributing guidelines](https://github.com/mggg/GerryChainJulia/blob/main/CONTRIBUTING.md).
```

File: docs/src/saving_results.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+serialize("example.jld", chain_data)
+```
+
+Then, in order to read back the saved data in another file, just run
+```
+chain_data = deserialize("example.jld")
+```
+Note that the `.jld` format is specific to Julia, so you won't be able to `deserialize` in a script written in another language, like Python or R.
+
+## Saving scores to csv
```
Comment (matthewkol186) [inline docs/src/saving_results.md:23]:
```suggestion
## Saving scores to CSV
```

File: docs/src/saving_results.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+**Resulting CSV**
+
+```
+cut_edges,vote_count_d_1,vote_count_d_2,vote_count_d_3,vote_share_d_1,vote_share_d_2,vote_share_d_3
+5,2,4,4,0.2,0.4,0.4
+6,3,5,2,0.3,0.5,0.2
+```
+
+## Saving score to JSON
```
Comment (matthewkol186) [inline docs/src/saving_results.md:46]:
```suggestion
## Saving scores to JSON
```

File: docs/src/scores.md [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+existing plan to the properties of the ensemble of plans generated by the chain.
+Thus, we need a way to record information about the plans generated during the
+progression of the chain. This is the point of `Score`s: each score has a
+scoring function that is evaluated on the current districting plan at each step
+of the chain. The resulting values are saved in a dictionary, which is itself
+added to a growing `ChainScoreData` object at each step of the chain.
+The `ChainScoreData` returned by `recom_chain()` and `flip_chain()` is like
+a "history" of the scores at each step of the chain.
+(See [get_scores_at_step](#get_scores_at_step) for information about how to
+retrieve the values of any subset of scores at a particular step in the chain.)
```
Comment (matthewkol186) [inline docs/src/scores.md:13]: Just checking - do these links still work in the new documentation?

File: src/scores.jl [hunk 6]
```diff
@@ -1,83 +1,105 @@
...
-        upon/modify some shared state.
-    """
     name::String
     scores::Array{S,1} where {S<:AbstractScore} # should be other AbstractScores
 end
 
 
+"""
+    ChainScoreData(scores::Array{S,1} where {S<:AbstractScore} # scores that were measured on a particular chain
+                   step_values::Array{Dict{String, Any}}) # array of Dicts which map {score name: score value}
```
Comment (matthewkol186) [inline src/scores.jl:64]: Is this the right format for this docstring? just checking

File: docs/src/accept.md [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+Pages   = ["accept.md"]
+```
+
+The following is a sample acceptance function included in the `GerryChain` library:
+
+```@docs
+always_accept
+```
+
+## Full Docs
```
Comment (matthewkol186) [inline docs/src/accept.md:33]: Can we change this phrase to "API" or something? "Full Docs" sounds a little awk to me.

**Comment by matthewkol186** [comment]
General note: the "Arguments" list here is a bit difficult to read. Any way we can change the format so it's more like a bulleted list or something?

image

**Review by matthewkol186 — APPROVED** [review]
This is SO good, @bsuwal ! The docs look so ✨official✨ and I love the easy-to-read layout. I agree with your point about having each argument be on a newline, although in cases where there's just two or three arguments, it's okay to have them all fit on one line (kinda like how in Python you only put each argument on its own line if having them all on one line would exceed some maximum line length). 

Also thanks for the very detailed PR summary! Beautifully done ✅ Giving you a Mega Slowking because I was so slow to review.

image

File: docs/src/index.md (hunk 1 above)
Comment (bsuwal) [reply docs/src/index.md:30]: you are right! fixed.

File: docs/src/scores.md (hunk 5 above)
Comment (bsuwal) [reply docs/src/scores.md:13]: great catch - they didn't! fixed.

File: src/scores.jl (hunk 6 above)
Comment (bsuwal) [reply src/scores.jl:64]: Great question... we actually do not expose this particular docstring in the API so I don't know - you can get `ChainScoreData` from `get_scores()` and `get_scores_at_step()` but we don't expect the user to ever instantiate it themself (should we? i think it is fine as is for now.)

File: docs/src/accept.md (hunk 7 above)
Comment (bsuwal) [reply docs/src/accept.md:33]: Sounds good! Done.

**Comment by bsuwal** [comment]
> General note: the "Arguments" list here is a bit difficult to read. Any way we can change the format so it's more like a bulleted list or something? Also, only some methods have a full explanation for each argument - seems fine for now, but we should probably aim for explanations for each argument in every function, right?
> 
> image

Fixed this!

Also you are right -- some methods have a full explanation for every argument but some don't  and that should be fixed - just opened up #86 so that we don't forget about it.

**Comment by bsuwal** [comment]
> This is SO good, @bsuwal ! The docs look so ✨official✨ and I love the easy-to-read layout. I agree with your point about having each argument be on a newline, although in cases where there's just two or three arguments, it's okay to have them all fit on one line (kinda like how in Python you only put each argument on its own line if having them all on one line would exceed some maximum line length).
> 
> Also thanks for the very detailed PR summary! Beautifully done ✅ Giving you a Mega Slowking because I was so slow to review.
> 
> image

I enjoy making these PRs so much because of your delightful reviews!!

**Pull request merged**